docs/standard/garbage-collection/implementing-dispose.md

```diff
@@ -1,158 +1,173 @@
+
+    > [!TIP]
+    > If you use the recommended technique discussed in the previous bullet, the <xref:System.Runtime.InteropServices.SafeHandle?displayProperty=nameWithType> class does this on your behalf.
+
+To help ensure that resources are always cleaned up appropriately, a <xref:System.IDisposable.Dispose%2A> method should be idempotent, such that it is callable multiple times without throwing an exception.
+
 The code example provided for the <xref:System.GC.KeepAlive%2A?displayProperty=nameWithType> method shows how garbage collection can cause a finalizer to run, while an unmanaged reference to the object or its members is still in use. It may make sense to utilize <xref:System.GC.KeepAlive%2A?displayProperty=nameWithType> to make the object ineligible for garbage collection from the start of the current routine to the point where this method is called.
-  
+
 <a name="Dispose2"></a>
-## Dispose() and Dispose(Boolean)  
-
-The <xref:System.IDisposable> interface requires the implementation of a single parameterless method, <xref:System.IDisposable.Dispose%2A>. However, the dispose pattern requires two `Dispose` methods to be implemented:  
-  
-- A public non-virtual (`NonInheritable` in Visual Basic) <xref:System.IDisposable.Dispose%2A?displayProperty=nameWithType> implementation that has no parameters.  
-  
-- A protected virtual (`Overridable` in Visual Basic) `Dispose` method whose signature is:  
-  
+## Dispose() and Dispose(bool)
+
+The <xref:System.IDisposable> interface requires the implementation of a single parameterless method, <xref:System.IDisposable.Dispose%2A>. However, the last two dispose pattern variants require two `Dispose` methods to be implemented:
+
+- A `public` non-virtual (`NonInheritable` in Visual Basic) <xref:System.IDisposable.Dispose%2A?displayProperty=nameWithType> implementation that has no parameters.
+
+- A `protected virtual` (`Overridable` in Visual Basic) `Dispose` method whose signature is:
+
   [!code-csharp[Conceptual.Disposable#8](../../../samples/snippets/csharp/VS_Snippets_CLR/conceptual.disposable/cs/dispose1.cs#8)]
-  [!code-vb[Conceptual.Disposable#8](../../../samples/snippets/visualbasic/VS_Snippets_CLR/conceptual.disposable/vb/dispose1.vb#8)]  
-  
-### The Dispose() overload
+  [!code-vb[Conceptual.Disposable#8](../../../samples/snippets/visualbasic/VS_Snippets_CLR/conceptual.disposable/vb/dispose1.vb#8)]
+
+  > [!NOTE]
+  > The `disposing` parameter should be `false` when called from a finalizer, and `true` when called from the <xref:System.IDisposable.Dispose%2A?displayProperty=nameWithType> method. In other words, it is `true` when deterministically called and `false` when non-deterministically called.
```

Comment:
To further complicate things, it's also false when called from DisposeAsync :)

Reply:
Do you have examples of this? Calling out `DisposeAsync` might be something that is out of scope for this iteration, but I'd love details and examples now - if you have them.

Reply:
DisposeAsync's pattern is based on the basic dispose pattern:

```C#
public async ValueTask DisposeAsync()
{
    await DisposeAsyncCore().ConfigureAwait(false);
    Dispose(false);
    GC.SuppressFinalize(this);
}

protected virtual ValueTask DisposeAsyncCore() { ... }
```

(where the presence or absence of `.ConfigureAwait(false)` is a topic unto itself)

* Code from the `if (disposing)` block(s) in `Dispose(bool)` gets copied to `DisposeAsyncCore()`, but switching to async variants (where applicable).
* `DisposeAsync` then calls the synchronous `Dispose(false)` to sweep out the remainder of the synchronous basic dispose pattern, e.g. any native resource cleanup code.
* Having swept both the (asynchronous variant of) the "disposing" and "not", `DisposeAsync` can now call GC.SuppressFinalize(this).

Based on the way that 99+% of `Dispose(bool)` methods it doesn't actually matter if `DisposeAsync` calls `Dispose(true)` or `Dispose(false)`, but we picked `false` to help show that things from the `false` branch didn't belong in `DisposeAsyncCore`.  As a practical detail, `Dispose(bool)` essentially ends up being the place to set a `_disposed = true`, since it needs to still run after `DisposeAsyncCore`.

Reply:
@bartonjs For possible new rules to add to https://github.com/DotNetAnalyzers/IDisposableAnalyzers, are guidelines for implemented DisposeAsync documented anywhere that we could reference?

E.g. should an analyzer check for the presence of .ConfigureAwait(false), or check that Dispose(false) is called if there is a Dispose(bool) method, etc?

Reply:
Hi @jnm2 - I'm working on capturing these guidelines next... See #18623

Reply:
That is awesome, thank you!